libr/bin/format/mach0/mach0.c

```diff
@@ -3311,92 +3321,89 @@ RSkipList *MACH0_(get_relocs)(struct MACH0_(obj_t) *bin) {
 										break;
 									}
 									*reloc = *ref;
-									reloc->addr = cursor;
+									reloc->addr = addr;
 									reloc->offset = paddr;
 									if (addend != -1) {
 										reloc->addend = addend;
 									}
 									r_skiplist_insert (relocs, reloc);
 								}
-								cursor += delta * wordsize;
+								addr += delta * wordsize;
 								if (!delta) {
 									break;
 								}
 							}
-						} else {
-							bprintf ("Error: BIND_SUBOPCODE_THREADED_APPLY with no current segment\n");
 						}
+						break;
+					default:
+						bprintf ("Error: Unexpected BIND_OPCODE_THREADED sub-opcode: 0x%x\n", imm);
 					}
 					break;
-				default:
-					bprintf ("Error: Unexpected BIND_OPCODE_THREADED sub-opcode: 0x%x\n", imm);
-				}
-				break;
-			}
-			case BIND_OPCODE_SET_DYLIB_ORDINAL_IMM:
-				lib_ord = imm;
-				break;
-			case BIND_OPCODE_SET_DYLIB_ORDINAL_ULEB:
-				lib_ord = ULEB();
-				break;
-			case BIND_OPCODE_SET_DYLIB_SPECIAL_IMM:
-				lib_ord = imm? (st8)(BIND_OPCODE_MASK | imm) : 0;
-				break;
-			case BIND_OPCODE_SET_SYMBOL_TRAILING_FLAGS_IMM: {
-				sym_name = (char*)ur.p;
-				while (*ur.p++ && ur.p<end) {
-					/* empty loop */
 				}
-				if (threaded_binds) {
+				case BIND_OPCODE_SET_DYLIB_ORDINAL_IMM:
+					lib_ord = imm;
 					break;
-				}
-				sym_ord = -1;
-				if (bin->symtab && bin->dysymtab.nundefsym < 0xffff) {
-					for (j = 0; j < bin->dysymtab.nundefsym; j++) {
-						int stridx = 0;
-						int iundefsym = bin->dysymtab.iundefsym;
-						if (iundefsym >= 0 && iundefsym < bin->nsymtab) {
-							int sidx = iundefsym + j;
-							if (sidx < 0 || sidx >= bin->nsymtab) {
-								continue;
+				case BIND_OPCODE_SET_DYLIB_ORDINAL_ULEB:
+					lib_ord = ULEB();
+					break;
+				case BIND_OPCODE_SET_DYLIB_SPECIAL_IMM:
+					lib_ord = imm? (st8)(BIND_OPCODE_MASK | imm) : 0;
+					break;
+				case BIND_OPCODE_SET_SYMBOL_TRAILING_FLAGS_IMM: {
+					sym_name = (char*)ur.p;
+					while (*ur.p++ && ur.p<end) {
+						/* empty loop */
+					}
+					if (threaded_binds) {
+						break;
+					}
+					sym_ord = -1;
+					if (bin->symtab && bin->dysymtab.nundefsym < 0xffff) {
+						for (j = 0; j < bin->dysymtab.nundefsym; j++) {
+							int stridx = 0;
```

Comment:
Note that I only changed the indentation level of this code to keep the scope as narrow as possible – but if we're changing these lines we might as well audit them 👌

```C
int stridx = 0;
int iundefsym = bin->dysymtab.iundefsym;
if (iundefsym >= 0 && iundefsym < bin->nsymtab) {
	int sidx = iundefsym + j;
	if (sidx < 0 || sidx >= bin->nsymtab) {
		continue;
	}
	stridx = bin->symtab[sidx].n_strx;
	if (stridx < 0 || stridx >= bin->symstrlen) {
		continue;
	}
}
if (!strcmp ((char *)bin->symstr + stridx, sym_name)) {
	sym_ord = j;
	break;
}
```

So `stridx` is declared and initialized to zero, but never read until it's assigned a new value here:

```C
stridx = bin->symtab[sidx].n_strx;
```

Here, `n_strx` is a `uint32_t`. And on the next line we check:

```C
	if (stridx < 0 || stridx >= bin->symstrlen) {
```

Where `stridx < 0` is impossible, so that should be removed. So we're left with this check:

```C
	if (stridx >= bin->symstrlen) {
```

Where `symstrlen` is an `int`. So ISTM we caan just drop the variable and inline that into the check:

```C
if (bin->symtab[sidx].n_strx >= bin->symstrlen) {
```

I also noticed that `iundefsym` is assigned to an `int` here, also coming from an `uint32_t`, and checking if it's `>= 0` (anything else is impossible), and used to compute `sidx`, also an `int` that cannot be `< 0`, but still there's a check for that.

Reply:
Ignore that, `ENOCOFFEE` so I totally misread that code 😂